hep/proposals/0001-decorators.md

```diff
@@ -0,0 +1,611 @@
+def fibonacci(fibo: FiboInput) -> FiboOutput:
+    # we will need to validate that users can only run/define certain kinds of expressions here
+    # and throw warnings if they are using syntax that is invalid
+    valid_number = hio.expr(fibo.num > 1)
+    previous_num = calculator(CalculatorInput(x=fibo.num, y=1, operation="sub"), when=valid_number)
+    previous_fibo = fibonacci(FiboInput(num=previous_num.result), when=valid_number)
+    current = calculator(CalculatorInput(x=fibo.num, y=previous_fibo.num), when=valid_number)
+    return FiboOutput(num=valid_number.check(current.num, 1))
+```
+
+### Parallel steps
+
+Converting the DAG code:
+
+```py
+from hera.workflows import WorkflowTemplate, parallel
+import hera.workflows.io as hio
+
+# We start by defining our Workflow Template
+wt = WorkflowTemplate(name="my-template")
+
+@wt.script  # Adds a new script template to the workflow template
+def setup() -> hio.Output:
+    return hio.Output(result="success")
+
+class ConcatInput(hio.Input):
+    word_a: str
+    word_b: str
+
+@wt.script  # Adds a new script template to the workflow template
+def concat(concat_input: ConcatInput) -> hio.Output:
+    return hio.Output(result=f"{concat_input.word_a} {concat_input.word_b}")
+
+class WorkerInput(hio.Input):
+    value_a: str
+    value_b: str
+
+class WorkerOutput(hio.Output):
+    value: str
+
+
+@wt.entrypoint
+@wt.steps
+def worker(worker_input: WorkerInput) -> WorkerOutput:
+    setup_step = setup()
+
+    # We can potentially add parameters such as `when` to `parallel` to add
+    # the `when` clause to all sub-steps
+    with parallel():
+        step_a = concat(ConcatInput(word_a=worker_input.value_a, word_b=setup_step.result))
+        step_b = concat(ConcatInput(word_a=worker_input.value_b, word_b=setup_step.result))
+
+    final_step = concat(ConcatInput(word_a=step_a.result, word_b=step_b.result))
+    return WorkerOutput(value=final_step.result)
+```
+
+## Container Template
+
+```python
+from hera.workflows import WorkflowTemplate, Container
+import hera.workflows.io as hio
+
+# We start by defining our Workflow Template
+wt = WorkflowTemplate(name="my-template")
+
+# This defines the template's inputs
+class MyInput(hio.Input):
+    user: str = "Hera"
+
+class MyOutput(hio.Output):
+    container_greeting: Annotated[
+        str,
+        Parameter(
+            name="container-greeting",
+            value_from={"path": "/tmp/hello_world.txt"},
+        ),
+    ]
+
+@wt.entrypoint
+@wt.container(command=["sh", "-c"], args=["echo {{inputs.parameters.user}} | tee /tmp/hello_world.txt"])
+def basic_hello_world(my_input: MyInput) -> hio.Output:
+    ...
+
+
+@wt.entrypoint
+@wt.container(command=["sh", "-c"])
+def advanced_hello_world(my_input: MyInput, template: Container) -> MyOutput:
+    # A 'MyOutput' object can be used to "mock" the output when running locally and allow us to
+    # inject outputs based on inputs.
+    output: MyOutput = MyOutput(my_input)
```

Comment:
This is invalid as it shouldn't be the user defined type taking in another one. This was originally a `hio.output` function.